Cancelation event not cached by server

### **Versions of relevant software used:**
@improbable-eng/grpc-web: v0.14.0 (go)
@improbable-eng/grpc-web: 0.14.0 (js)

### **What happened:**
Transport: Websocket.
When user makes `unary` or `server stream` request and cancels it, the cancelation event 
isn't cached by the server.

### **What you expected to happen:**
Server should cach the cancelation event.

### **How to reproduce it (as minimally and precisely as possible):**
Transport: Websocket.
Create `unary` or `server stream` service, cancel it on user side and try caching this cancel event on server.

### **My fix:**
webSocketWrappedReader in websocket_wrapper.go  should read one message ahead from websocket.
I've already implemented this in code, so if you would like to I could make a PR.

That's because tests never run in CI, see details in CI logs:
```
+ karma start ./karma.conf.ts --single-run
21.023633 testserver: 2021/03/10 21:56:10 Starting servers. http1.1 port: 9090, http1.1 empty port: 9095, http2 port: 9100, http2 empty port: 9105
23.214382 
23.214444 START:
23.238775 10 03 2021 21:56:12.631:INFO [karma-server]: Karma v6.0.3 server started at https://testhost:9876/
23.238950 10 03 2021 21:56:12.632:INFO [launcher]: Launching browsers Linux_chrome_41 with concurrency 4
23.249055 10 03 2021 21:56:12.641:INFO [launcher]: Starting browser Linux_chrome_41
27.561711 10 03 2021 21:56:16.955:INFO [Chrome 41.0.2272.76 (Linux i686)]: Connected on socket j5sxrhlc67zaPrVIAAAB with id 40222341
28.575147 Chrome 41.0.2272.76 (Linux i686) ERROR
28.575223   An error was thrown in afterAll
28.575240   Uncaught SyntaxError: Unexpected token )
28.575251   SyntaxError: Unexpected token )
28.578474 
28.578532 Finished in 0.107 secs / 0 secs @ 21:56:17 GMT+0000 (Coordinated Universal Time)
28.578542 
28.578618 SUMMARY:
28.578703 ✔ 0 tests completed
28.579476 
28.579533 ✖ Error while running the tests! Exit code: 1
28.579542 
28.579766 Chrome 41.0.2272.76 (Linux i686) ERROR
28.579776   Some of your tests did a full page reload!
28.580078 
28.580124 ✖ Error while running the tests! Exit code: 1
28.580134 
+ killGoTestServer
+ echo 'Killing testserver...'
+ killall testserver
31.919059 Killing testserver...
+ killTunnels
+ echo 'Killing Sauce Labs Tunnels...'
+ [[ -z '' ]]
+ kill 1204
+ [[ ! -z '' ]]
31.922051 Killing Sauce Labs Tunnels...
31.922156 10 Mar 21:56:21 - Got signal terminated
npm ERR! code 1
npm ERR! path /go/src/github.com/improbable-eng/grpc-web/integration_test
npm ERR! command failed
npm ERR! command sh -c ./run-with-tunnel.sh ./run-with-testserver.sh karma start ./karma.conf.ts --single-run

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/circleci/.npm/_logs/2021-03-10T21_56_21_322Z-debug.log
32.571899 10 Mar 21:56:21 - Cleaning up the tunnel connection to maki1639.miso.saucelabs.com
32.571969 10 Mar 21:56:21 - Cleaning up.
32.572073 10 Mar 21:56:21 - Removing tunnel 7f29d07de32649f7b4791d1b7c52af21.
32.572118 10 Mar 21:56:21 - Waiting for any active jobs using this tunnel to finish.
32.572167 10 Mar 21:56:21 - Press CTRL-C again to shut down immediately.
32.572206 10 Mar 21:56:21 - Note: if you do this, tests that are still running will fail.
36.607968 10 Mar 21:56:26 - All jobs using the tunnel have finished.
36.608046 10 Mar 21:56:26 - Waiting for the connection to terminate...
37.608364 10 Mar 21:56:27 - Connection closed (8).
37.608428 10 Mar 21:56:27 - Goodbye.
```
I think CI is broken and needs a fix. If you would checkout on my branch and run tests locally you would see them failing.

--- a/go/grpcweb/websocket_wrapper.go
+++ b/go/grpcweb/websocket_wrapper.go
@@ -176,6 +176,16 @@ func (w *webSocketWrappedReader) Read(p []byte) (int, error) {
 
 	// If the frame consists of only a single byte of value 1 then this indicates the client has finished sending
 	if len(framePayload) == 1 && framePayload[0] == 1 {
+		go func() {
+			for {
+				messageType, _, err := w.wsConn.Read(w.context)
+				if err == io.EOF || messageType == 0 {
+					// The client has closed the connection. Indicate to the response writer that it should close
+					w.cancel()
+					return
+				}
+			}
+		}()
 		return 0, io.EOF
 	}
 

--- a/integration_test/ts/src/cancellation.spec.ts
+++ b/integration_test/ts/src/cancellation.spec.ts
@@ -14,7 +14,7 @@ import {
 } from "../_proto/improbable/grpcweb/test/test_pb";
 import {TestService, TestUtilService} from "../_proto/improbable/grpcweb/test/test_pb_service";
 import {DEBUG, continueStream} from "./util";
-import { runWithHttp1AndHttp2 } from "./testRpcCombinations";
+import { runWithHttp1AndHttp2, runWithSupportedTransports } from "./testRpcCombinations";
 
 describe("Cancellation", () => {
   runWithHttp1AndHttp2(({testHostUrl}) => {
@@ -52,91 +52,94 @@ describe("Cancellation", () => {
       assert.equal(transportCancelFuncInvoked, true, "transport's cancel func must be invoked");
     });
 
-    it("should handle aborting a streaming response mid-stream with propagation of the disconnection to the server", (done) => {
-      let onMessageId = 0;
+    runWithSupportedTransports((transport) => {
+      it("should handle aborting a streaming response mid-stream with propagation of the disconnection to the server", (done) => {
+        let onMessageId = 0;
 
-      const streamIdentifier = `rpc-${Math.random()}`;
+        const streamIdentifier = `rpc-${Math.random()}`;
 
-      const ping = new PingRequest();
-      ping.setValue("hello world");
-      ping.setResponseCount(100); // Request more messages than the client will accept before cancelling
-      ping.setStreamIdentifier(streamIdentifier);
+        const ping = new PingRequest();
+        ping.setValue("hello world");
+        ping.setResponseCount(100); // Request more messages than the client will accept before cancelling
+        ping.setStreamIdentifier(streamIdentifier);
 
-      let reqObj: grpc.Request;
+        let reqObj: grpc.Request;
 
-      // Checks are performed every 1s = 15s total wait
-      const maxAbortChecks = 15;
+        // Checks are performed every 1s = 15s total wait
+        const maxAbortChecks = 15;
 
-      const numMessagesBeforeAbort = 5;
+        const numMessagesBeforeAbort = 5;
 
-      const doAbort = () => {
-        DEBUG && debug("doAbort");
-        reqObj.close();
+        const doAbort = () => {
+          DEBUG && debug("doAbort");
+          reqObj.close();
 
-        // To ensure that the transport is successfully closing the connection, poll the server every 1s until
-        // it confirms the connection was closed. Connection closure is immediate in some browser/transport combinations,
-        // but can take several seconds in others.
-        function checkAbort(attempt: number) {
-          DEBUG && debug("checkAbort", attempt);
-          continueStream(testHostUrl, streamIdentifier, (status) => {
-            DEBUG && debug("checkAbort.continueStream.status", status);
+          // To ensure that the transport is successfully closing the connection, poll the server every 1s until
+          // it confirms the connection was closed. Connection closure is immediate in some browser/transport combinations,
+          // but can take several seconds in others.
+          function checkAbort(attempt: number) {
+            DEBUG && debug("checkAbort", attempt);
+            continueStream(testHostUrl, streamIdentifier, (status) => {
+              DEBUG && debug("checkAbort.continueStream.status", status);
 
-            const checkStreamClosedRequest = new CheckStreamClosedRequest();
-            checkStreamClosedRequest.setStreamIdentifier(streamIdentifier);
-            grpc.unary(TestUtilService.CheckStreamClosed, {
-              debug: DEBUG,
-              request: checkStreamClosedRequest,
-              host: testHostUrl,
-              onEnd: ({message}) => {
-                const closed = ( message as CheckStreamClosedResponse ).getClosed();
-                DEBUG && debug("closed", closed);
-                if (closed) {
-                  done();
-                } else {
-                  if (attempt >= maxAbortChecks) {
-                    assert.ok(closed, `server did not observe connection closure within ${maxAbortChecks} seconds`);
+              const checkStreamClosedRequest = new CheckStreamClosedRequest();
+              checkStreamClosedRequest.setStreamIdentifier(streamIdentifier);
+              grpc.unary(TestUtilService.CheckStreamClosed, {
+                debug: DEBUG,
+                request: checkStreamClosedRequest,
+                host: testHostUrl,
+                onEnd: ({message}) => {
+                  const closed = ( message as CheckStreamClosedResponse ).getClosed();
+                  DEBUG && debug("closed", closed);
+                  if (closed) {
                     done();
                   } else {
-                    setTimeout(() => {
-                      checkAbort(attempt + 1);
-                    }, 1000);
+                    if (attempt >= maxAbortChecks) {
+                      assert.ok(closed, `server did not observe connection closure within ${maxAbortChecks} seconds`);
+                      done();
+                    } else {
+                      setTimeout(() => {
+                        checkAbort(attempt + 1);
+                      }, 1000);
+                    }
                   }
-                }
-              },
-            })
-          });
-        }
-
-        checkAbort(0);
-      };
-
-      reqObj = grpc.invoke(TestService.PingList, {
-        debug: DEBUG,
-        request: ping,
-        host: testHostUrl,
-        onHeaders: (headers: grpc.Metadata) => {
-          DEBUG && debug("headers", headers);
-        },
-        onMessage: (message: PingResponse) => {
-          assert.ok(message instanceof PingResponse);
-          DEBUG && debug("onMessage.message.getCounter()", message.getCounter());
-          assert.strictEqual(message.getCounter(), onMessageId++);
-          if (message.getCounter() === numMessagesBeforeAbort) {
-            // Abort after receiving numMessagesBeforeAbort messages
-            doAbort();
-          } else if (message.getCounter() < numMessagesBeforeAbort) {
-            // Only request the next message if not yet aborted
-            continueStream(testHostUrl, streamIdentifier, (status) => {
-              DEBUG && debug("onMessage.continueStream.status", status);
+                },
+              })
             });
           }
-        },
-        onEnd: (status: grpc.Code, statusMessage: string, trailers: grpc.Metadata) => {
-          DEBUG && debug("status", status, "statusMessage", statusMessage, "trailers", trailers);
-          // onEnd shouldn't be called if abort is called prior to the response ending
-          assert.fail();
-        }
-      });
-    }, 20000);
+
+          checkAbort(0);
+        };
+
+        reqObj = grpc.invoke(TestService.PingList, {
+          debug: DEBUG,
+          request: ping,
+          host: testHostUrl,
+          transport: transport,
+          onHeaders: (headers: grpc.Metadata) => {
+            DEBUG && debug("headers", headers);
+          },
+          onMessage: (message: PingResponse) => {
+            assert.ok(message instanceof PingResponse);
+            DEBUG && debug("onMessage.message.getCounter()", message.getCounter());
+            assert.strictEqual(message.getCounter(), onMessageId++);
+            if (message.getCounter() === numMessagesBeforeAbort) {
+              // Abort after receiving numMessagesBeforeAbort messages
+              doAbort();
+            } else if (message.getCounter() < numMessagesBeforeAbort) {
+              // Only request the next message if not yet aborted
+              continueStream(testHostUrl, streamIdentifier, (status) => {
+                DEBUG && debug("onMessage.continueStream.status", status);
+              });
+            }
+          },
+          onEnd: (status: grpc.Code, statusMessage: string, trailers: grpc.Metadata) => {
+            DEBUG && debug("status", status, "statusMessage", statusMessage, "trailers", trailers);
+            // onEnd shouldn't be called if abort is called prior to the response ending
+            assert.fail();
+          }
+        });
+      }, 20000);
+    })
   });
 });

--- a/integration_test/ts/src/client.websocket.spec.ts
+++ b/integration_test/ts/src/client.websocket.spec.ts
@@ -4,9 +4,14 @@ import { grpc } from "@improbable-eng/grpc-web";
 import { debug } from "../../../client/grpc-web/src/debug";
 import { assert } from "chai";
 // Generated Test Classes
-import { PingRequest, PingResponse } from "../_proto/improbable/grpcweb/test/test_pb";
-import { TestService } from "../_proto/improbable/grpcweb/test/test_pb_service";
-import { DEBUG, DISABLE_WEBSOCKET_TESTS } from "./util";
+import {
+  CheckStreamClosedRequest,
+  CheckStreamClosedResponse,
+  PingRequest,
+  PingResponse
+} from "../_proto/improbable/grpcweb/test/test_pb";
+import { TestService, TestUtilService } from "../_proto/improbable/grpcweb/test/test_pb_service";
+import { continueStream, DEBUG, DISABLE_WEBSOCKET_TESTS } from "./util";
 import { headerTrailerCombos, runWithHttp1AndHttp2 } from "./testRpcCombinations";
 
 if (DISABLE_WEBSOCKET_TESTS) {
@@ -73,7 +78,7 @@ if (DISABLE_WEBSOCKET_TESTS) {
 
       describe("bidirectional (websockets)", () => {
         headerTrailerCombos((withHeaders, withTrailers) => {
-          it("should make a bidirectional request that is terminated by the client", (done) => {
+          it("should make a bidirectional request that is ended by the client finishing sending", (done) => {
             let didGetOnHeaders = false;
             let counter = 1;
             let lastMessage = `helloworld:${counter}`;
@@ -121,6 +126,102 @@ if (DISABLE_WEBSOCKET_TESTS) {
 
               assert.equal(counter, 10, "counter should have been incremented to 10");
               done();
+            });
+            client.start();
+
+            // send initial message
+            client.send(ping);
+          });
+        });
+
+        headerTrailerCombos((withHeaders, withTrailers) => {
+          it("should make a bidirectional request that is aborted by the client with propagation of the disconnection to the server", (done) => {
+            let didGetOnHeaders = false;
+            let counter = 1;
+            const streamIdentifier = `rpc-${Math.random()}`;
+            let lastMessage = `helloworld:${counter}`;
+            const ping = new PingRequest();
+            ping.setStreamIdentifier(streamIdentifier);
+            ping.setSendHeaders(withHeaders);
+            ping.setSendTrailers(withTrailers);
+            ping.setValue(lastMessage);
+
+            const client = grpc.client(TestService.PingPongBidi, {
+              debug: DEBUG,
+              host: testHostUrl,
+              transport: grpc.WebsocketTransport(),
+            });
+
+            // Checks are performed every 1s = 15s total wait
+            const maxAbortChecks = 15;
+
+            const doAbort = () => {
+              DEBUG && debug("doAbort");
+              client.close();
+
+              // To ensure that the transport is successfully closing the connection, poll the server every 1s until
+              // it confirms the connection was closed. Connection closure is immediate in some browser/transport combinations,
+              // but can take several seconds in others.
+              function checkAbort(attempt: number) {
+                DEBUG && debug("checkAbort", attempt);
+                continueStream(testHostUrl, streamIdentifier, (status) => {
+                  DEBUG && debug("checkAbort.continueStream.status", status);
+
+                  const checkStreamClosedRequest = new CheckStreamClosedRequest();
+                  checkStreamClosedRequest.setStreamIdentifier(streamIdentifier);
+                  grpc.unary(TestUtilService.CheckStreamClosed, {
+                    debug: DEBUG,
+                    request: checkStreamClosedRequest,
+                    host: testHostUrl,
+                    onEnd: ({message}) => {
+                      const closed = ( message as CheckStreamClosedResponse ).getClosed();
+                      DEBUG && debug("closed", closed);
+                      if (closed) {
+                        done();
+                      } else {
+                        if (attempt >= maxAbortChecks) {
+                          assert.ok(closed, `server did not observe connection closure within ${maxAbortChecks} seconds`);
+                          done();
+                        } else {
+                          setTimeout(() => {
+                            checkAbort(attempt + 1);
+                          }, 1000);
+                        }
+                      }
+                    },
+                  })
+                });
+              }
+
+              checkAbort(0);
+            };
+
+            client.onHeaders((headers: grpc.Metadata) => {
+              DEBUG && debug("headers", headers);
+              didGetOnHeaders = true;
+              if (withHeaders) {
+                assert.deepEqual(headers.get("HeaderTestKey1"), ["ServerValue1"]);
+                assert.deepEqual(headers.get("HeaderTestKey2"), ["ServerValue2"]);
+              }
+            });
+            client.onMessage((message: PingResponse) => {
+              assert.ok(message instanceof PingResponse);
+              assert.deepEqual(message.getValue(), lastMessage);
+
+              if (counter === 10) {
+                doAbort();
+              } else {
+                counter++;
+                lastMessage = `helloworld:${counter}`;
+                const ping = new PingRequest();
+                ping.setValue(lastMessage);
+                client.send(ping);
+              }
+            });
+            client.onEnd((status: grpc.Code, statusMessage: string, trailers: grpc.Metadata) => {
+              DEBUG && debug("status", status, "statusMessage", statusMessage);
+              // onEnd shouldn't be called if abort is called prior to the response ending
+              assert.fail();
             });
             client.start();
 

